nibabies/workflows/bold/alignment.py

```diff
@@ -0,0 +1,310 @@
+        (dilate, resample, [("out_file", "in_file")]),
+        (create_label, resample, [("out_file", "template")]),
+        (resample, smooth, [("out_file", "in_file")]),
+        (smooth, separate, [("out_file", "in_file")]),
+        # end loop
+        (mul_roi, fmt_agg_rois, [("out_file", "rois")]),
+        (fmt_agg_rois, agg_rois, [
+            ("first_image", "in_file"),
+            ("op_files", "operand_files"),
+            ("op_string", "op_string")]),
+        (separate, combine_rois, [("volume_all_file", "in_files")]),
+        (combine_rois, outputnode, [("out", "subcortical_file")]),
+    ])
+    # fmt: on
+    return workflow
+
+
+def parse_roi_labels(label_file):
+    """
+    Parse a label file composed of one or more sets of:
+    <labelname>
+    <key> <red> <green> <blue> <alpha>
+
+    Return a list of structure names and label keys.
+
+    >>> structs, ids = parse_roi_labels(test_data / "labelfile.txt")
+    >>> structs
+    ['CEREBELLUM_LEFT', 'THALAMUS_LEFT', 'CAUDATE_LEFT']
+    >>> ids
+    [8, 10, 11]
+    """
+
+    with open(label_file) as fp:
+        lines = fp.readlines()
+    if len(lines) % 2 == 1:
+        raise RuntimeError("Label file is incomplete or invalid")
+    structs, label_ids = [], []
+    for idx, line in enumerate(lines):
+        if idx % 2 == 0:
+            structs.append(line.strip())
+        else:
+            label_ids.append(int(line.split(' ', 1)[0]))
+    return structs, label_ids
+
+
+def format_volume_rois(structs, rois):
+    """Format volume arguments for CiftiCreateDenseFromTemplate."""
+    return [(struct, roi) for struct, roi in zip(structs, rois)]
+
+
+def format_agg_rois(rois):
+    """
+    Helper function to format MultiImageMaths command.
+
+    Parameters
+    ----------
+    rois : `list` of `str`s
+        List of files
+
+    Returns
+    -------
+    first_image
+    op_files
+    op_string
+
+    """
+    return rois[0], rois[1:], ("-add %s " * (len(rois) - 1)).strip()
+
+
+def merge_rois(in_files):
+    """
+    Aggregate individual 4D ROI files together into a single subcortical NIfTI.
+    All ROI images are sanity checked with regards to:
+    1) Shape
+    2) Affine
+    3) Overlap
+
+    If any of these checks fail, an ``AssertionError`` will be raised.
+    """
+    import os
+    import nibabel as nb
+    import numpy as np
+
+    data = None
+    nonzero = 0
+    shape = None
+    for roi in in_files:
+        img = nb.load(roi)
+        roi_data = np.asanyarray(img.dataobj)
+        roi_nonzero = np.count_nonzero(roi_data)
+        if data is None:
+            # set first ROI as template
+            shape = img.shape
+            affine = img.affine
+            data = roi_data
+            nonzero += roi_nonzero
+            continue
+        assert img.shape == shape, "Mismatch in image shape"
+        assert np.allclose(img.affine, affine), "Mismatch in affine"
+        nonzero += roi_nonzero
```

Comment:
I find it a bit clearer to start with the first element outside the loop and then loop over the rest than to special-case the first time through a loop.

```suggestion
    img = nb.load(in_files[0])
    data = np.array(img.dataobj)
    affine = img.affine
    header = img.header
    
    nonzero = np.any(data, axis=3)
    for roi in in_files[1:]:
        img = nb.load(roi)
        assert img.shape == data.shape, "Mismatch in image shape"
        assert np.allclose(img.affine, affine), "Mismatch in affine"
        roi_data = np.asanyarray(img.dataobj)
        roi_nonzero = np.any(roi_data, axis=3)
        assert not np.any(roi_nonzero & nonzero), "Overlapping ROIs"
        nonzero |= roi_nonzero
```

Also:

* Saving the header
* Switching to using `data.shape` rather than keeping a separate shape tuple
* Checking shape and affine before loading the data
* Creating a boolean volume to directly check overlap, rather than checking the count at the end.